### PR TITLE
Add Smart Contracts Docs

### PR DESCRIPTION
**Pull request by fabioberger**
This PR:
- Adds smart contract docs to website
  - Convert Doxity docs JSON to docs agnostic JSON
  - Refactor docs pages and components to support solidity-specific concepts
  - Add introduction markdown to top of smart contract docs pointing to wiki
  - Show deployed network badges next to each contract header
  - Display event types
- Add smart contract docs to menus
- Generalized versions stored in redux so they can be used by 0x.js and contract docs.

screen shot 2017-10-04 at 14 53 57

### CODE REVIEW

File: ts/pages/documentation/smart_contracts_documentation.tsx [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+        const versionDocObj = await docUtils.getJSONDocFileAsync(versionFileNameToFetch, DOC_JSON_ROOT);
+        const docAgnosticFormat = doxityUtils.convertToDocAgnosticFormat(versionDocObj as DoxityDocObj);
+
+        this.setState({
+            docAgnosticFormat,
+        }, () => {
+            this.scrollToHash();
+        });
+    }
+}
```
Comment (LogvinovLeon) [inline ts/pages/documentation/smart_contracts_documentation.tsx]: 400 lines?
Refactor that to different components.
I feel like if smth has a render*() method - this is a clear sign, that it should be separated.

File: ts/pages/home/home.tsx [hunk 2]
```diff
@@ -112,7 +112,7 @@ export class Home extends React.Component<HomeProps, HomeState> {
                                 <div className="flex sm-hide xs-hide">
                                     <Link to="/docs/0xjs">
                                         <RaisedButton
-                                            label="Docs"
+                                            label="0x.js"
```
Comment (LogvinovLeon) [inline ts/pages/home/home.tsx]: 0x.js Docs

File: ts/pages/home/home.tsx [hunk 3]
```diff
@@ -126,6 +126,11 @@ export class Home extends React.Component<HomeProps, HomeState> {
                                             label="Whitepaper"
                                         />
                                     </a>
+                                    <Link to="/docs/contracts">
+                                        <FlatButton
+                                            label="Contracts"
```
Comment (LogvinovLeon) [inline ts/pages/home/home.tsx]: Smart Contrats Docs

File: ts/types.ts [hunk 4]
```diff
@@ -378,25 +378,43 @@ export interface DocSection {
...
     name: string;
-    comment?: string;
     returnComment?: string|undefined;
-    source: Source;
     callPath: string;
     parameters: Parameter[];
     returnType: Type;
+    comment?: string;
+
+    // Only Typescript methods
```
Comment (LogvinovLeon) [inline ts/types.ts]: What about having three different types?
Method, SolidityMethod, TypeScriptMethod?

File: ts/utils/constants.ts [hunk 5]
```diff
@@ -108,11 +118,17 @@ export const constants = {
         'TRST': '/images/token_icons/trust.png',
     } as {[symbol: string]: string},
     networkNameById: {
-        1: 'Frontier',
-        3: 'Ropsten',
-        4: 'Rinkeby',
-        42: 'Kovan',
+        1: Networks.mainnet,
```
Comment (LogvinovLeon) [inline ts/utils/constants.ts]: 👍

File: ts/utils/doxity_utils.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    TypeDocTypes,
+    EventArg,
+} from 'ts/types';
+
+export const doxityUtils = {
+    convertToDocAgnosticFormat(doxityDocObj: DoxityDocObj): DocAgnosticFormat {
+        const docAgnosticFormat: DocAgnosticFormat = {};
+        _.each(doxityDocObj, (doxityContractObj: DoxityContractObj, contractName: string) => {
+            const doxityConstructor = _.find(doxityContractObj.abiDocs, (abiDoc: DoxityAbiDoc) => {
+                return abiDoc.type === 'constructor';
```
Comment (LogvinovLeon) [inline ts/utils/doxity_utils.ts]: Maybe create an enum?

File: ts/utils/doxity_utils.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                };
+                constructors.push(constructor);
+            }
+
+            const doxityMethods = _.filter(doxityContractObj.abiDocs, (abiDoc: DoxityAbiDoc) => {
+                return this._isMethod(abiDoc);
+            });
+            const methods = _.map(doxityMethods, (doxityMethod: DoxityAbiDoc) => {
+                const outputIfExists = !_.isUndefined(doxityMethod.outputs) ?
+                    doxityMethod.outputs[0] :
```
Comment (LogvinovLeon) [inline ts/utils/doxity_utils.ts]: This is unclear.
Why only the first one?
If there is only one - why is it an array?

File: ts/utils/doxity_utils.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                    undefined;
+                const returnTypeIfExists = !_.isUndefined(outputIfExists) ?
+                    this._convertType(outputIfExists.type) :
+                    undefined;
+                const callPath = contractName !== 'ZRXToken' ?
+                    `${contractName[0].toLowerCase()}${contractName.slice(1)}.` :
+                    `${contractName.slice(0, 3).toLowerCase()}${contractName.slice(3)}.`;
```
Comment (LogvinovLeon) [inline ts/utils/doxity_utils.ts]: 3 is a magic number

File: ts/utils/doxity_utils.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                return method;
+            });
+
+            const doxityProperties = _.filter(doxityContractObj.abiDocs, (abiDoc: DoxityAbiDoc) => {
+                return this._isProperty(abiDoc);
+            });
+            const properties = _.map(doxityProperties, (doxityProperty: DoxityAbiDoc) => {
+                let typeName = doxityProperty.outputs[0].type;
+                if (!_.isEmpty(doxityProperty.inputs)) {
+                    typeName = `(${doxityProperty.inputs[0].type} => ${typeName})`;
```
Comment (LogvinovLeon) [inline ts/utils/doxity_utils.ts]: 0 is a magic number

File: ts/pages/home/home.tsx (hunk 3 above)
Comment (fabioberger) [reply ts/pages/home/home.tsx]: `Smart Contracts`

File: ts/pages/documentation/smart_contracts_documentation.tsx (hunk 1 above)
Comment (fabioberger) [reply ts/pages/documentation/smart_contracts_documentation.tsx]: Agreed but this refactor will be a separate PR.

File: ts/utils/doxity_utils.ts (hunk 7 above)
Comment (fabioberger) [reply ts/utils/doxity_utils.ts]: Added a comment.

File: ts/utils/doxity_utils.ts (hunk 8 above)
Comment (fabioberger) [reply ts/utils/doxity_utils.ts]: Yes, added a comment.

File: ts/utils/doxity_utils.ts (hunk 9 above)
Comment (fabioberger) [reply ts/utils/doxity_utils.ts]: Added a comment

File: ts/types.ts [hunk 10]
```diff
@@ -408,11 +408,17 @@ export interface Method {
     returnType: Type;
     comment?: string;
 
-    // Only Typescript methods
+    // Only Solidity methods
+    isConstant?: boolean;
+    isPayable?: boolean;
```
Comment (LogvinovLeon) [inline ts/types.ts]: shit

File: ts/utils/doxity_utils.ts [hunk 11]
```diff
@@ -129,7 +134,7 @@ export const doxityUtils = {
         return isMethod;
     },
     _isProperty(abiDoc: DoxityAbiDoc) {
-        if (abiDoc.type !== 'function') {
+        if (abiDoc.type !== DoxityAbiDocType.Function) {
```
Comment (LogvinovLeon) [inline ts/utils/doxity_utils.ts]: Add this to web3 typings

**Pull request merged**